### ISSUE TITLE
Is there a quick way to add Inheritence to VBA?

### ISSUE THREAD

**Issue by Beakerboy** [NaturalDocs/NaturalDocs#28]
If a VBA .cls file has a line starting with "Implements", then it is implementing an interface. Is there an easy way to make NaturalDocs aware of this by modifying one of the text configuration files, or is this something that would have to be compiled into the code?  Any suggestions on where to look so I can do it myself and supply a patch?

**Comment by NaturalDocs** [comment]
My understanding of VBA is that it gets the class name from the file name and it's not declared in the file contents, correct?  Then there's no way to do it without editing the code.  The generalized parser could work with something like this:

    ' Class: MyClass
    class MyClass implements MyInterface

but if it can't find the name in the comment anywhere in the prototype it assumes they don't go together.

    ' Class: MyClass
    Implements MyInterface

If you want to take a stab at changing that the code to match them is in Engine/Source/Languages/Languages.Parsing.cs, function AddBasicPrototype.  Then you may also need to change ParseClassPrototype in the same file to interpret it correctly, because that function's also not expecting that syntax.

**Comment by Beakerboy** [comment]
Here’s the full header of an exported class module:

````vba
VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = "SQLSelect"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = True
Option Explicit

' Class: SQLSelect
' A SQL Select query
Implements iSQLQuery
````
So the name of the class is a part of a properly expected class.

**Comment by NaturalDocs** [comment]
Right.  Natural Docs can't deal with that as it is right now.  I can look into adding special code for VBA to handle this in the next version.

If it implements multiple interfaces, a quick google shows me they're on separate lines and not on one, so

    Implements iSQLQuery
    Implements iSomethingElse

and not

    Implements iSQLQuery, iSomethingElse

correct?  Would they also always appear one immediately after another like this, or can they be scattered throughout the file like

    Implements iSQLQuery
    ... code for this interface ...

    Implements iSomethingElse
    ... code for this interface ...

**Comment by Beakerboy** [comment]
If you have an interface called `iFoo` that defined a function called `Bar(sBaz as String)`, the child Class will define the function as `iFoo_Bar(sBaz as String)`. The function definitions can be in any order. This means if a class implements multiple interfaces where each interface defines a function with name `Bar`, they have to be defined separately in the child, one as `iFoo_bar` and the other as `iQix_Bar`

**Comment by NaturalDocs** [comment]
I'm only concerned with the Implements statements since that's what Natural Docs needs to detect.  If they always appear together I can have the code look for all the consecutive Implements statements appearing right after the Class comment, which fits into the current architecture well.  If they don't always appear together I have to search the entire file, which is probably doable but harder without full language support.

**Comment by Beakerboy** [comment]
Oh, I see...I don't know if the `Implements` statements have to be all together. I've never had to Implement multiple interfaces.

**Comment by Beakerboy** [comment]
Alternatively, instead of changing the tokenization and prototype, would it be easier to add something to the comment to force NaturalDocs to know there's a parent interface. I imagine there are languages which have function and class declarations much different than any of the prototypes are looking for. In those cases, wouldn't it be nice if all the details could be specified in the comment block?

// Class: MyClass
// Does something interesting with a number
// Extends: MyParent
// Implements: MyInterface
// Parameters:
//    x - the important integer
//
// Returns: an integer

**Comment by sancarn** [comment]
@NaturalDocs Implements statements do not have to be declared together, but cannot be declared under a `End Sub`, `End Property` or `End Function` statement.

```vb
Implements Class2

Const foo = 1
Private Declare Function Pinapple Lib "something" () As String
Private Enum something
    arr
End Enum
Private Type someStruct
    x As Integer
End Type

Implements Class3


Public Sub thing()

End Sub
Public Function thingy()

End Function
Public Property Get x() As String

End Property
```

Also it might be worth seperating your implementations into seperate implementations for VB.NET and VB6/VBA. In VB.NET you can define classes and can define.

VB6/VBA has the class header as follows:

```vb
VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = "SQLSelect"      '<----- this is the name of the class
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute VB_Exposed = True

implements A
 '...
implements B
 '...
'Sub/Prop/Function definitions
```

VB.NET has class header as follows:

```vb
Class SomeName
  Implements A, B
 '...
End Class
```
